pkgdb/src/buildenv/realise.cc

```diff
@@ -60,59 +56,397 @@ namespace flox::buildenv {
 #  error "COMMON_NIXPKGS_URL must be set to a locked flakeref of nixpkgs to use"
 #endif
 
-#ifndef FLOX_BASH_BIN
-#  error "FLOX_BASH_BIN must be set to the path of a Bash executable"
+#ifndef FLOX_BASH_PKG
+#  error "FLOX_BASH_PKG must be set to the path of the nix bash package"
+#endif
+
+#ifndef FLOX_COREUTILS_PKG
+#  error \
+    "FLOX_COREUTILS_PKG must be set to the path of the nix coreutils package"
+#endif
+
+#ifndef FLOX_GNUSED_PKG
+#  error "FLOX_GNUSED_PKG must be set to the path of the nix gnused package"
+#endif
+
+#ifndef FLOX_PROCPS_PKG
+#  error "FLOX_PROCPS_PKG must be set to the path of the nix procps package"
+#endif
+
+#ifndef FLOX_CACERT_PKG
+#  error "FLOX_CACERT_PKG must be set to the path of the nixpkgs cacert package"
+#endif
+
+#ifdef linux
+#  ifndef FLOX_LOCALE_ARCHIVE
+#    error "FLOX_LOCALE_ARCHIVE_PKG must be set to the LOCALE_ARCHIVE variable"
+#  endif
+#else  // darwin
+#  ifndef FLOX_PATH_LOCALE
+#    error "FLOX_PATH_LOCALE_PKG must be set to the PATH_LOCALE variable"
+#  endif
+#  ifndef FLOX_NIX_COREFOUNDATION_RPATH
+#    error \
+      "FLOX_NIX_COREFOUNDATION_RPATH must be set to the NIX_COREFOUNDATION_RPATH variable"
+#  endif
 #endif
 
 /* -------------------------------------------------------------------------- */
 
+// Top-level activate script, always invoked with nix bash.
+const char * const ACTIVATE_SCRIPT = R"_(
+# Flox environment activation script.
+[ "${_FLOX_PKGDB_VERBOSITY:-0}" -eq 0 ] || set -x
+
+# TODO: add getopt arg parser for following args:
+# -c "<cmd> <args>": specify exact command args to pass to shell
+# --turbo: invoke commands directly without involving userShell
+# --noprofile: do not source `[profile]` scripts
+
+# Set FLOX_ENV as the path by which all flox scripts can make reference to
+# the environment to which they belong. Use this to define the path to the
+# activation scripts directory.
+# TODO: reconcile with CLI which should be setting this. We must override
+#       the value coming from the CLI for now because it won't be set for
+#       container invocations, and it would have the incorrect value for
+#       nested flox activations.
+_FLOX_ENV="$( $_coreutils/bin/dirname -- "${BASH_SOURCE[0]}" )"
+if [ -n "$FLOX_ENV" -a "$FLOX_ENV" != "$_FLOX_ENV" ]; then
+  echo "WARN: detected change in FLOX_ENV: $FLOX_ENV -> $_FLOX_ENV" >&2
+fi
+export FLOX_ENV="$_FLOX_ENV"
+
+# The rust CLI contains sophisticated logic to set $FLOX_SHELL based on the
+# process listening on STDOUT, but that won't happen when activating from
+# the top-level activation script, so fall back to $SHELL as a default.
+FLOX_SHELL="${FLOX_SHELL:-$SHELL}"
+
+# Set all other variables derived from FLOX_ENV. We previously did this
```

Comment:
Thanks for the comments - they distill down to these two points:

- concern about moving activation logic from the flox CLI to the environment itself
    - that decision was really made for us when we observed that containerized flox environments could not afford to carry the ~200MB closure penalty of including a copy of flox in every environment
    - we made an attempt to jury-rig activations by simulating the final action of the CLI to launch the user shell, but that obviously introduces a UX disparity as new features are introduced to the CLI over time
    - therefore in order to achieve feature parity between flox-activated and containerized renderings of environments we really had no choice but to move the activation logic into the environment itself
    - the good news, and the thing that confirmed for me that this was the correct design decision, is that the dockerfile configuration for activating flox environments got much, much simpler with this change as we replaced all the "sh -c" malarkey with a single "flox activate"
- concern over the use of bash
    - this will be automation that users will be very interested to "lift the hood" and inspect at runtime, so its essential that it be interpreted at runtime, and bash is undoubtedly the best-established and universally-{accepted,understood} scripting language for startup/service orchestration scripts
    - again, closure size is an issue, because whatever interpreter we use will need to be shipped in the closure
    - Python would be a good alternative, but even the closure size for python3Minimal comes back at the ~33Mb mark as opposed to bash's ~3Mb
    - the etc-profiles mechanism is already bash-based, so the top-level activate script as implemented in this PR is only slightly increasing, rather than introducing a new exposure to bash

So with that I'd suggest that we draw this conversation to a close, get it out there and then look out for ongoing user sentiment in response to the use of bash in both the activation script and etc-profiles. Believe me I would love to be programming in a better language than bash, and in particular Python holds if we could get

Reply:
Adding this thread to the followup ticket https://github.com/flox/flox/issues/1343